### PR TITLE
issue/2662 Formalised plugin, data, schema and translate API

### LINKED ISSUE #2662: Formalise Adapt framework, plugin, data and schema API
### Subject of the issue
Currently the data structures, schemas which represent them, the way the plugins extend the structures and the way these pieces function in Adapt Framework must be inferred from the grunt tasks.

We should formalise these behaviours in code, such that they are easier to understand, critique and develop.

A boundary can be created between the file data on disk (JSON, schema, package and bower JSON) and the compilation processes which utilise this data.

### PR DESCRIPTION
**Pull request by oliverfoster**
fixes #2662 
#### Additions
* Added [grunt/helpers folder](https://github.com/adaptlearning/adapt_framework/tree/issue/2662/grunt/helpers) with APIs for Framework, Plugins, Schemas, Data and Translate
* Made all tasks JSON file agnostic (except config.json which is necessarily separate)
* `grunt translate:import` is extended with a new `--test=true` parameter.
* Added [fs-extra](https://github.com/jprichardson/node-fs-extra), [globs](https://github.com/stephenmathieson-boneyard/node-globs#readme) and [lodash](https://lodash.com/) to the package.json
* Added [manifest creation API](https://github.com/adaptlearning/adapt_framework/blob/issue/2662/grunt/helpers/data/Language.js#L110-L130) for #2645 

#### Changes
* Changed grunt/config/build-config and grunt/config/replace to use the new API
* Updated grunt/helpers to use new API, fixed some linting and escaping issues, changed the module.export to work with intellisense and treated core as a plugin in the includes regular expression
* Fixed linting and escaping issues in grunt/tasks/handlebars (so that the grunt folder passes linting)
* Changed grunt/tasks/schema-defaults, grunt/tasks/tracking-insert, grunt/tasks/tracking-remove, grunt/tasks/tracking-reset and grunt/tasks/translate to use the new API, [simplifying them greatly](https://github.com/adaptlearning/adapt_framework/blob/issue/2662/grunt/tasks/tracking-insert.js)

#### Removed
* Deleted grunt/config/tracking-insert, grunt/config/tracking-remove, grunt/config/tracking-reset and grunt/config/translate configs as they are no longer needed
* Deleted previous translate code as it is now [encapsulated in the new API](https://github.com/adaptlearning/adapt_framework/blob/issue/2662/grunt/helpers/Translate.js)

#### Notes
The build tasks should carry on working as normal, this change should be non-breaking but should increase code clarity dramatically. It is necessary to formalise these in order to the move to v6.
Intellisense should work really nicely for the new API in at least VSCode.
If you want to debug grunt you can use [grunt-brk](https://www.npmjs.com/package/grunt-brk) and [chrome://inspect](chrome://inspect).

### CODE REVIEW

File: grunt/helpers/Framework.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+  /**
+   * Returns a Data instance for either the src/course or build/course folder
+   * depending on the specification of the useOutputData property on either the
+   * function or the Framework instance.
+   * @returns {Data}
+   */
+  getData({ useOutputData = this.useOutputData } = {}) {
+    const data = useOutputData
+      ? new Data({
```
Comment (tomgreenfield) [inline grunt/helpers/Framework.js]: Suggest moving these ternary operators up a line.

File: grunt/helpers/Framework.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+   * @param {string} options.rootPath
+   * @param {string} options.outputPath
+   * @param {string} options.sourcePath
+   * @param {function} options.includedFilter
+   * @param {string} options.jsonext
+   * @param {boolean} options.useOutputData
+   * @param {function} options.log
+   * @param {function} options.warn
+   */
+  constructor({ rootPath = process.cwd(), outputPath = process.cwd() + '/build/', sourcePath = process.cwd() + '/src/', includedFilter = function() { return true; }, jsonext = 'json', useOutputData = false, log = console.log, warn = console.warn } = {}) {
```
Comment (tomgreenfield) [inline grunt/helpers/Framework.js]: I'd personally format these on new lines for readability:
```js
constructor({
  rootPath = process.cwd(),
  outputPath = process.cwd() + '/build/',
  sourcePath = process.cwd() + '/src/',
  includedFilter = function() { return true; },
  jsonext = 'json',
  useOutputData = false,
  log = console.log,
  warn = console.warn
} = {}) {
```

File: grunt/helpers/Translate.js [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+      components: 5
+    };
+    exportTextData.sort((a, b) => {
+      const typeSort = ((typeSortLevel[a.file] || 100) - (typeSortLevel[b.file] || 100));
+      return typeSort || a.id.length - b.id.length || a.id.localeCompare(b.id);
+    });
+
+    // output based upon format options
+    const outputFolder = path.join(this.languagePath, this.masterLang);
+    if (!fs.existsSync(outputFolder)) {
```
Comment (tomgreenfield) [inline grunt/helpers/Translate.js]: Do we need the `if` statement?

File: grunt/helpers/data/Language.js [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    }, {});
+
+    this.getCourseFileItem();
+
+    return this;
+  }
+
+  /** @type {Boolean} */
+  get hasChanged() {
+    return this.files.find(file => file.hasChanged);
```
Comment (tomgreenfield) [inline grunt/helpers/data/Language.js]: `.some()`?

File: grunt/helpers/data/Language.js [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+      }
+      return index;
+    }, {});
+
+    this.getCourseFileItem();
+
+    return this;
+  }
+
+  /** @type {Boolean} */
```
Comment (tomgreenfield) [inline grunt/helpers/data/Language.js]: Decap `Boolean`?

File: grunt/helpers/lib/JSONFile.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      err.number = 10013;
+      throw err;
+    }
+
+    return this;
+  }
+
+  /**
+   * This is useful for files such as config.json or course.json which only have
+   * on item/object per file.
```
Comment (tomgreenfield) [inline grunt/helpers/lib/JSONFile.js]: Typo: `on`.

File: grunt/helpers/lib/JSONFile.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+   * @returns {JSONFileItem}
+   */
+  get firstFileItem() {
+    return this.fileItems && this.fileItems[0];
+  }
+
+  /**
+   * Marks this file as having changed. This should be called after changing an
```
Comment (tomgreenfield) [inline grunt/helpers/lib/JSONFile.js]: Typo?

File: grunt/helpers/plugins/Plugin.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    if (this.name === 'core' || config.hasOwnProperty('component')) {
+      return 'component';
+    } else if (config.hasOwnProperty('extension')) {
+      return 'extension';
+    } else if (config.hasOwnProperty('menu')) {
+      return 'menu';
+    } else if (config.hasOwnProperty('theme')) {
+      return 'theme';
+    }
+    throw new Error('Unknown plugin type');
```
Comment (tomgreenfield) [inline grunt/helpers/plugins/Plugin.js]: Don't necessarily need all the `else if`s; could do something like the below?
```js
if (this.name === 'core') return 'component';
const type = Object.keys(this.packageJSONFile.firstFileItem.item).find(key => {
  return [ 'component', 'extension', 'menu', 'theme' ].includes(key);
});
if (!type) throw new Error('Unknown plugin type');
return type;
```

File: grunt/helpers/Schemas.js [hunk 9]
```diff
@@ -0,0 +1,199 @@
+const _ = require('lodash');
```
Comment (tomgreenfield) [inline grunt/helpers/Schemas.js]: Does this need to added to the package JSON?

File: grunt/helpers/plugins/Plugin.js (hunk 8 above)
Comment (oliverfoster) [reply grunt/helpers/plugins/Plugin.js]: I've added a few more variables to make it a bit clearer.

File: grunt/helpers/Framework.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    return data;
+  }
+
+  /** @returns {Plugins} */
+  getPlugins({
+    includedFilter = this.includedFilter
+  } = {}) {
+    const plugins = new Plugins({
+      framework: this.framework,
+      includedFilter: includedFilter,
```
Comment (jamesrea83) [inline grunt/helpers/Framework.js]: assignment of includedFilter can be destructured

File: grunt/helpers/Translate.js [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+  }
+
+  /**
+   * Produces a single JSON file or a series of CSV files representing all of the
+   * files and file items in the data structure.
+   * @returns {Translate}
+   */
+  async export() {
+
+    const schemas = new Schemas({ framework: this.framework, includedFilter: this.includedFilter, sourcePath: this.sourcePath });
```
Comment (jamesrea83) [inline grunt/helpers/Translate.js]: bit more destructuring could be done here-

`new Schemas({ framework, includedFilter, sourcePath } = this)`

Comment (oliverfoster) [reply grunt/helpers/Translate.js]: I did think about this, and figured it was perhaps a little too concise and bleedy for my tastes, although perfectly valid and concise.

Comment (oliverfoster) [reply grunt/helpers/Translate.js]: Tom G has a genius solution, but it's not clear what's going on at a glance:
```js
(({ a, b, c }) => ({ a, b, c }))({ a: 1, b: 2, c: 3, d: 4 })
```
```js
functionReturnValue = ArrowFunction((argsPickDestructorVariableCreation) => return(newObjectAssignmentFromVariables))CallArrowFunction(input)
```

File: grunt/helpers/schema/Schema.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+class Schema {
+
+  /**
+   * @param {Object} options
+   * @param {Framework} options.framework
+   * @param {string} options.name
+   * @param {Object} options.json
+   * @param {string} options.filePath
+   * @param {string) options.globalsType
```
Comment (lc-alexanderbenesch) [inline grunt/helpers/schema/Schema.js]: Typo

**Comment by oliverfoster** [comment]
Awaiting a framework release and a few more reviews before merging.

**Comment by oliverfoster** [comment]
This pr is ready for +1s if anyone has the inclination.

File: grunt/tasks/tracking-remove.js [hunk 13]
```diff
@@ -1,34 +1,8 @@
...
-    var sourcedir = isOutputDir ? grunt.option('outputdir') : grunt.config('sourcedir');
-
-    var blocksFiles = grunt.file.expand(path.join(sourcedir, options.blocksFile));
-    var courseFiles = grunt.file.expand(path.join(sourcedir, options.courseFile));
-
-    for (var i = 0; i < blocksFiles.length; i++) {
-      removeTrackingIds(blocksFiles[i], courseFiles[i]);
-    }
-
+    const data = Helpers.getFramework().getData();
```
Comment (moloko) [inline grunt/tasks/tracking-remove.js]: I know it's not part of your changes but I think the check for spoor in L4 should be removed.

Comment (oliverfoster) [reply grunt/tasks/tracking-remove.js]: Will have a crack at this when I layout moving tracking ids from blocks to components.

**Pull request merged**